smithy-aws-cloudformation-traits/src/main/java/software/amazon/smithy/aws/cloudformation/traits/ResourceIndex.java

```diff
@@ -0,0 +1,479 @@
+                    resource.getUpdate().ifPresent(operationId -> {
+                        operationIndex.getInput(operationId).ifPresent(input -> {
+                            updatePropertyMutabilities(resourceId, operationId, input,
+                                    SetUtils.of(Mutability.WRITE), this::addWriteMutability);
+                        });
+                    });
+
+                    // Apply any members found through the trait's additionalSchemas property.
+                    for (ShapeId additionalSchema : pair.getRight().getAdditionalSchemas()) {
+                        StructureShape shape = model.expectShape(additionalSchema, StructureShape.class);
+                        updatePropertyMutabilities(resourceId, null, shape,
+                                SetUtils.of(), Function.identity());
+                    }
+                });
+    }
+
+    public static ResourceIndex of(Model model) {
+        return model.getKnowledge(ResourceIndex.class, ResourceIndex::new);
+    }
+
+    /**
+     * Get all members of the CloudFormation resource.
+     *
+     * @param resource ShapeID of a resource.
+     * @return Returns all members that map to CloudFormation resource
+     *   properties.
+     */
+    public Map<String, ResourcePropertyDefinition> getProperties(ToShapeId resource) {
+        return resourcePropertyMutabilities.getOrDefault(resource.toShapeId(), MapUtils.of())
+                .entrySet().stream()
+                .filter(entry -> !getExcludedProperties(resource).contains(entry.getValue().getShapeId()))
+                .collect(MapUtils.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    /**
+     * Gets the specified member of the CloudFormation resource.
+     *
+     * @param resource ShapeID of a resource
+     * @param propertyName Name of the property to retrieve
+     * @return The property definition.
+     */
+    public Optional<ResourcePropertyDefinition> getProperty(ToShapeId resource, String propertyName) {
+        return Optional.ofNullable(getProperties(resource).get(propertyName));
+    }
+
+    /**
+     * Get create-specifiable-only members of the CloudFormation resource.
+     *
+     * These properties can be specified only during resource creation and
+     * can be returned in a `read` or `list` request.
+     *
+     * @param resource ShapeID of a resource.
+     * @return Returns create-only member names that map to CloudFormation resource
+     *   properties.
+     */
+    public Set<String> getCreateOnlyProperties(ToShapeId resource) {
+        return getConstrainedProperties(resource, definition -> {
+            Set<Mutability> mutabilities = definition.getMutabilities();
+            return mutabilities.contains(Mutability.CREATE) && !mutabilities.contains(Mutability.WRITE);
+        });
+    }
+
+    /**
+     * Get read-only members of the CloudFormation resource.
+     *
+     * These properties can be returned by a `read` or `list` request,
```

Comment:
Nit: these are javadoc comments and need stuff like `{@code read}` instead of markdown backticks.